launch_testing_ros/launch_testing_ros/wait_for_topics.py

```diff
@@ -120,13 +136,16 @@ def start_subscribers(self, topic_tuples):
                     10
                 )
             )
+            # Initialize ring buffer of received messages
+            self.received_messages_buffer[topic_name] = deque(maxlen=self.max_number_of_messages)
 
     def callback_template(self, topic_name):
 
         def topic_callback(data):
             if topic_name not in self.received_topics:
                 self.get_logger().debug('Message received for ' + topic_name)
                 self.received_topics.add(topic_name)
+                self.received_messages_buffer[topic_name].append(data)
```

Comment:
Isn't this only going to append data the first time the topic is encountered?

Reply:
I'm kind of confused why we need a `deque` for each topic if we only ever push one message into it.

Reply:
Naively I would expect that only one message is received, too. However, when running the example included with the source code, the behavior that I observe is that multiple messages are published. I am attaching here a portion of the output of the `wait_for_topic_launch_test.py` script as evidence.

One reason is that, from the moment the first message is received and `msg_event_object` is set, until the `wait` method is exited, there might be other messages received due to the asynchronous nature of ROS2 nodes.

Another reason is that, in the example, the "talker" node is publishing messages periodically without any way to synchronize with the "listener" node. Ideally, the talker node should publish only one or more messages only when the "listener" node is ready to receive them. I try solving this problem in my other pull request #356.

In conclusion, I think @adityapande-1995 suggestion to use a `deque` object is good because it can be used to check edge cases when multiple messages are received and generally is more future-proof.

```sh
# launch_test /root/ros2_humble/install/launch_testing_ros/share/launch_testing_ros/examples/wait_for_topic_launch_test.py
[INFO] [launch]: All log files can be found below /root/.ros/log/2023-03-10-07-21-43-728296-748e7bd0d623-2534
[INFO] [launch]: Default logging verbosity is set to INFO
test_topics_successful (wait_for_topic_launch_test.TestFixture)
All the supplied topics should be read successfully. ... [INFO] [python3-1]: process started with pid [2565]
[INFO] [python3-2]: process started with pid [2567]
[INFO] [python3-3]: process started with pid [2569]
[INFO] [python3-4]: process started with pid [2571]
[INFO] [python3-5]: process started with pid [2573]
[python3-2] [INFO] [1678432904.672566992] [demo_node_1]: Publishing: "Hello World: 0"
[python3-3] [INFO] [1678432904.685455303] [demo_node_2]: Publishing: "Hello World: 0"
[python3-4] [INFO] [1678432904.700987787] [demo_node_3]: Publishing: "Hello World: 0"
[python3-1] [INFO] [1678432904.702492445] [demo_node_0]: Publishing: "Hello World: 0"
[python3-5] [INFO] [1678432904.708634882] [demo_node_4]: Publishing: "Hello World: 0"
[python3-2] [INFO] [1678432904.765118265] [demo_node_1]: Publishing: "Hello World: 1"
[python3-3] [INFO] [1678432904.776693934] [demo_node_2]: Publishing: "Hello World: 1"
[python3-4] [INFO] [1678432904.792138703] [demo_node_3]: Publishing: "Hello World: 1"
[python3-1] [INFO] [1678432904.793491690] [demo_node_0]: Publishing: "Hello World: 1"
[python3-5] [INFO] [1678432904.799117885] [demo_node_4]: Publishing: "Hello World: 1"
[python3-2] [INFO] [1678432904.865036169] [demo_node_1]: Publishing: "Hello World: 2"
[python3-3] [INFO] [1678432904.876631619] [demo_node_2]: Publishing: "Hello World: 2"
[python3-4] [INFO] [1678432904.892106463] [demo_node_3]: Publishing: "Hello World: 2"
[python3-1] [INFO] [1678432904.893564367] [demo_node_0]: Publishing: "Hello World: 2"
[python3-5] [INFO] [1678432904.899091952] [demo_node_4]: Publishing: "Hello World: 2"
[python3-2] [INFO] [1678432904.965028530] [demo_node_1]: Publishing: "Hello World: 3"
[python3-3] [INFO] [1678432904.976670809] [demo_node_2]: Publishing: "Hello World: 3"
[python3-4] [INFO] [1678432904.991995366] [demo_node_3]: Publishing: "Hello World: 3"
[python3-1] [INFO] [1678432904.993493504] [demo_node_0]: Publishing: "Hello World: 3"
ok
test_topics_unsuccessful (wait_for_topic_launch_test.TestFixture)
```

Reply:
> Isn't this only going to append data the first time the topic is encountered?

Yes, that is right. I did not notice that. Thank you for pointing it out.

Reply:
Disclaimer: the following reply is not directly related to this comment. 

Even if the bug you pointed out is fixed, naively I would expect that only one message is received, too. However, when running the example included with the source code, the behavior that I observe is that multiple messages are published. I am attaching here a portion of the output of the `wait_for_topic_launch_test.py` script as evidence.

One reason is that, from the moment the first message is received and `msg_event_object` is set, until the `wait` method is exited, there might be other messages received due to the asynchronous nature of ROS2 nodes.

Another reason is that, in the example, the "talker" node is publishing messages periodically without any way to synchronize with the "listener" node. Ideally, the talker node should publish one or more messages only when the "listener" node is ready to receive them. I try solving this problem in my other pull request #356.

In conclusion, I think @adityapande-1995 suggestion to use a `deque` object is good because it can check edge cases when multiple messages are received, and generally is more future-proof.

```sh
# launch_test /root/ros2_humble/install/launch_testing_ros/share/launch_testing_ros/examples/wait_for_topic_launch_test.py
[INFO] [launch]: All log files can be found below /root/.ros/log/2023-03-10-07-21-43-728296-748e7bd0d623-2534
[INFO] [launch]: Default logging verbosity is set to INFO
test_topics_successful (wait_for_topic_launch_test.TestFixture)
All the supplied topics should be read successfully. ... [INFO] [python3-1]: process started with pid [2565]
[INFO] [python3-2]: process started with pid [2567]
[INFO] [python3-3]: process started with pid [2569]
[INFO] [python3-4]: process started with pid [2571]
[INFO] [python3-5]: process started with pid [2573]
[python3-2] [INFO] [1678432904.672566992] [demo_node_1]: Publishing: "Hello World: 0"
[python3-3] [INFO] [1678432904.685455303] [demo_node_2]: Publishing: "Hello World: 0"
[python3-4] [INFO] [1678432904.700987787] [demo_node_3]: Publishing: "Hello World: 0"
[python3-1] [INFO] [1678432904.702492445] [demo_node_0]: Publishing: "Hello World: 0"
[python3-5] [INFO] [1678432904.708634882] [demo_node_4]: Publishing: "Hello World: 0"
[python3-2] [INFO] [1678432904.765118265] [demo_node_1]: Publishing: "Hello World: 1"
[python3-3] [INFO] [1678432904.776693934] [demo_node_2]: Publishing: "Hello World: 1"
[python3-4] [INFO] [1678432904.792138703] [demo_node_3]: Publishing: "Hello World: 1"
[python3-1] [INFO] [1678432904.793491690] [demo_node_0]: Publishing: "Hello World: 1"
[python3-5] [INFO] [1678432904.799117885] [demo_node_4]: Publishing: "Hello World: 1"
[python3-2] [INFO] [1678432904.865036169] [demo_node_1]: Publishing: "Hello World: 2"
[python3-3] [INFO] [1678432904.876631619] [demo_node_2]: Publishing: "Hello World: 2"
[python3-4] [INFO] [1678432904.892106463] [demo_node_3]: Publishing: "Hello World: 2"
[python3-1] [INFO] [1678432904.893564367] [demo_node_0]: Publishing: "Hello World: 2"
[python3-5] [INFO] [1678432904.899091952] [demo_node_4]: Publishing: "Hello World: 2"
[python3-2] [INFO] [1678432904.965028530] [demo_node_1]: Publishing: "Hello World: 3"
[python3-3] [INFO] [1678432904.976670809] [demo_node_2]: Publishing: "Hello World: 3"
[python3-4] [INFO] [1678432904.991995366] [demo_node_3]: Publishing: "Hello World: 3"
[python3-1] [INFO] [1678432904.993493504] [demo_node_0]: Publishing: "Hello World: 3"
ok
test_topics_unsuccessful (wait_for_topic_launch_test.TestFixture)
```

Reply:
Fixed in commit 7c9e181fb3c4a3257f5315bfe217a4748019f0c3